### PR TITLE
[PLUGIN-1528] Add support to specify scale and rounding-mode for decimal datatype in set-type directive

### PR DESCRIPTION
**Pull request by vanathi-g**
Currently, there is no way to set the precision and scale of converted decimal values when using the set-type directive. 
## Changes
- Add `scale` and `rounding-mode` as optional arguments to the set-type directive
- `rounding-mode` maps to the different rounding modes available in Java and is used when the decimal value's scale is not equal to the set scale.
- By default, no value for scale is assumed (current behaviour) and `round-mode` = HALF_EVEN is assumed when scale is set.
- Modified unit test that checks conversion to decimal type and directive parsing. Added necessary new tests.

Note: Support to specify expected precision will be added later.

### CODE REVIEW

File: wrangler-core/src/main/java/io/cdap/directives/column/SetType.java
Comment (sau42shri) [inline wrangler-core/src/main/java/io/cdap/directives/column/SetType.java]: Although this is ok for smaller set of arguments. In this case, you are searching the arguments twice for "scale", since both contains and value will search for the input. Instead, you can simply take the value in a param and use that for null (similar to TryGetValue you have in dictionaries). You can choose to ignore this comment, but would be good to put this into practice for any such search pattern.

Comment (iam-divya) [reply wrangler-core/src/main/java/io/cdap/directives/column/SetType.java]: Arguments is implemented as a map here. 'Contains' map to Map's containsKey(https://github.com/data-integrations/wrangler/blob/8be2f88adc4fb69fb8c103301228d32f35bdea96/wrangler-core/src/main/java/io/cdap/wrangler/parser/MapArguments.java#L153) and value maps to map's get function which does not invoke containsKey again.

File: wrangler-core/src/main/java/io/cdap/wrangler/utils/ColumnConverter.java
Comment (iam-divya) [inline wrangler-core/src/main/java/io/cdap/wrangler/utils/ColumnConverter.java]: Why is this line needed? Flow from UI(not command line), these new args wont be set and do we throw an exception in that case?

File: wrangler-core/src/main/java/io/cdap/wrangler/utils/ColumnConverter.java
Comment (iam-divya) [inline wrangler-core/src/main/java/io/cdap/wrangler/utils/ColumnConverter.java]: nit: Do you want to call the variable something else? like convertedObject or something else instead of 'decimal'?

File: wrangler-core/src/test/java/io/cdap/directives/column/SetTypeTest.java
Comment (iam-divya) [inline wrangler-core/src/test/java/io/cdap/directives/column/SetTypeTest.java]: We should also be supporting just, 'set-type str_col decimal' without any scale or rounding args

Comment (vanathi-g) [reply wrangler-core/src/test/java/io/cdap/directives/column/SetTypeTest.java]: yes, that is supported

File: wrangler-core/src/main/java/io/cdap/directives/column/SetType.java
Comment (tivv) [inline wrangler-core/src/main/java/io/cdap/directives/column/SetType.java:67]: Why do we need both arguments for rounding? It looks more complex than having 1. 
Also, we should allow to set precision as well, should not we?

File: wrangler-core/src/main/java/io/cdap/directives/column/SetType.java
Comment (tivv) [inline wrangler-core/src/main/java/io/cdap/directives/column/SetType.java]: By default we should do exactly as we did before. We can't change the existing logic

File: wrangler-core/src/main/java/io/cdap/directives/column/SetType.java
Comment (vanathi-g) [reply wrangler-core/src/main/java/io/cdap/directives/column/SetType.java:67]: 1. Two arguments for rounding because if the user isn't familiar with Java rounding modes, they can just enable rounding and we can assume HALF_UP. 
2. We can't set a 'max' precision on a Java BigDecimal object. Since Wrangler infers datatypes for the output schema from the actual row values, we can't propagate the set precision to the output schema.

File: wrangler-core/src/main/java/io/cdap/directives/column/SetType.java
Comment (vanathi-g) [reply wrangler-core/src/main/java/io/cdap/directives/column/SetType.java]: This is the default scale used when rounding is enabled but scale is not specified. The existing logic is not changed.

**Comment by iam-divya** [comment]
LGTM. Please wait for Vitalii's LGTM before merging.

**Review by iam-divya — APPROVED** [review]

File: wrangler-core/src/main/java/io/cdap/directives/column/SetType.java
Comment (tivv) [reply wrangler-core/src/main/java/io/cdap/directives/column/SetType.java:67]: I don't think introducing two parameter makes it simpler. Any assumptions is source of confusion. And it's not java-specific, you can find more or less same set everywhere, starting from https://en.wikipedia.org/wiki/Rounding

File: wrangler-core/src/main/java/io/cdap/directives/column/SetType.java
Comment (tivv) [reply wrangler-core/src/main/java/io/cdap/directives/column/SetType.java]: I believe now it just convert to BigDecimal. E.g. currently if you have `1.23456` value it will remain `1.23456` value and would work if you set schema properly. With this change it's going to fail as default scale is 3 and  RoundingMode.UNNECESSARY.

**Review by tivv — CHANGES_REQUESTED** [review]
Please see comments inline.

File: wrangler-core/src/main/java/io/cdap/directives/column/SetType.java
Comment (vanathi-g) [reply wrangler-core/src/main/java/io/cdap/directives/column/SetType.java:67]: Done, removed the 'round' argument

File: wrangler-core/src/main/java/io/cdap/directives/column/SetType.java
Comment (vanathi-g) [reply wrangler-core/src/main/java/io/cdap/directives/column/SetType.java]: Changed the code to preserve current logic.

File: wrangler-core/src/main/java/io/cdap/directives/column/SetType.java
Comment (tivv) [inline wrangler-core/src/main/java/io/cdap/directives/column/SetType.java]: This is actually where we can have a default value of, say, https://docs.oracle.com/javase/7/docs/api/java/math/RoundingMode.html#HALF_EVEN. This would make it same as floating point rounding and users would be able to use "settype decimal, 2" without rounding mode with reasonal default and without errors

File: wrangler-core/src/main/java/io/cdap/directives/datamodel/DataModelMapColumn.java
Comment (tivv) [inline wrangler-core/src/main/java/io/cdap/directives/datamodel/DataModelMapColumn.java:155]: Do you really need to change it? If so, add a test. I am not sure what `args.contains("scale")` would return and if you would get an NPE.
BTW: It would be nice to handle null scale as well same as no scale

File: wrangler-core/src/main/java/io/cdap/wrangler/utils/ColumnConverter.java
Comment (tivv) [inline wrangler-core/src/main/java/io/cdap/wrangler/utils/ColumnConverter.java]: Please add a proper error message similar to line 291

**Review by tivv — COMMENTED** [review]
Please also update all documentation (`.md`) files where the directive is refered

File: wrangler-core/src/main/java/io/cdap/directives/datamodel/DataModelMapColumn.java
Comment (vanathi-g) [reply wrangler-core/src/main/java/io/cdap/directives/datamodel/DataModelMapColumn.java:155]: Setting it as null is alright because convertType is a utility method where setting scale as null is handled anyways. In any case, args.contains("scale") returns false when scale is null (ex, when scale isn't specified in the directive). Null scale is the same as no scale.

File: wrangler-core/src/main/java/io/cdap/directives/column/SetType.java
Comment (vanathi-g) [reply wrangler-core/src/main/java/io/cdap/directives/column/SetType.java]: Why HALF_EVEN? Initially I assumed the default rounding mode as HALF_UP but the idea is to keep the current behaviour (no rounding if the scale is different) the same right?

File: wrangler-core/src/main/java/io/cdap/wrangler/utils/ColumnConverter.java
Comment (vanathi-g) [reply wrangler-core/src/main/java/io/cdap/wrangler/utils/ColumnConverter.java]: Done

File: wrangler-core/src/main/java/io/cdap/directives/column/SetType.java
Comment (tivv) [reply wrangler-core/src/main/java/io/cdap/directives/column/SetType.java]: According to the documentation HALF_EVEN is the one used in IEEE-754. Also it has better math properties (statistically neutral). 

Yes, we should keep the current behavior unchanged - when scale is not set, we should do as before. Now if scale is set, we are free to provide a good default for the rounding mode as this is new behavior

Comment (tivv) [reply wrangler-core/src/main/java/io/cdap/directives/column/SetType.java]: But for the default rounding mode it would be great to have references on what other systems do

Comment (vanathi-g) [reply wrangler-core/src/main/java/io/cdap/directives/column/SetType.java]: Makes sense. Done, changed the default mode to HALF_EVEN.

File: wrangler-core/src/main/java/io/cdap/directives/column/SetType.java
Comment (tivv) [inline wrangler-core/src/main/java/io/cdap/directives/column/SetType.java:60]: nit: double empty line

File: wrangler-core/src/main/java/io/cdap/directives/column/SetType.java
Comment (tivv) [inline wrangler-core/src/main/java/io/cdap/directives/column/SetType.java]: Add the value you are parsing to the message and add `e` as cause

File: wrangler-core/src/main/java/io/cdap/wrangler/parser/MigrateToV2.java
Comment (tivv) [inline wrangler-core/src/main/java/io/cdap/wrangler/parser/MigrateToV2.java]: Adjust docs to the latest revision

File: wrangler-core/src/main/java/io/cdap/wrangler/utils/ColumnConverter.java
Comment (tivv) [inline wrangler-core/src/main/java/io/cdap/wrangler/utils/ColumnConverter.java]: Add `e` as a cause.  Also I don't think the message is telling the story. The most often case would be using RoundingMode.UNNESSESARY and the rounding is needed. It's better to say something about the value and scale tried, e.g. that scale can't be set to X for value Y with rounding mode Z

**Review by tivv — COMMENTED** [review]
Last polishing, generally looks good. 
Please also update all documentation files (*.md) that refer to the directive (fast search shows 5 files with the directive).

**Comment by tivv** [comment]
And add the JIRA reference to the title

File: wrangler-core/src/main/java/io/cdap/directives/column/SetType.java
Comment (vanathi-g) [reply wrangler-core/src/main/java/io/cdap/directives/column/SetType.java]: Done.

File: wrangler-core/src/main/java/io/cdap/wrangler/parser/MigrateToV2.java
Comment (vanathi-g) [reply wrangler-core/src/main/java/io/cdap/wrangler/parser/MigrateToV2.java]: Done, changed the necessary readme files too.

File: wrangler-core/src/main/java/io/cdap/wrangler/utils/ColumnConverter.java
Comment (vanathi-g) [reply wrangler-core/src/main/java/io/cdap/wrangler/utils/ColumnConverter.java]: Done, added a better error message

**Review by tivv — APPROVED** [review]

**Pull request merged**
